File: partiql-plan/src/main/java/org/partiql/plan/rex/RexStruct.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;
import org.partiql.spi.types.PType;

import java.util.List;

/**
 * Logical struct expression abstract base class.
 */
public abstract class RexStruct extends RexBase {

    /**
     * @return new RexStruct instance
     */
    @NotNull
    public static RexStruct create(@NotNull List<Field> fields) {
        return new Impl(fields);
    }

    /**
     * @return a field constructor instance
     */
    @NotNull
    public static Field field(Rex key, Rex value) {
        return new Field(key, value);
    }

    /**
     * @return list of struct fields (NOT operands)
     */
    @NotNull
    public abstract List<Field> getFields();

    @NotNull
    @Override
    protected final RexType type() {
        return RexType.of(PType.struct());
    }

    @NotNull
    @Override
    protected List<Operand> operands() {
        return List.of();
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitStruct(this, ctx);
    }

    /**
     * Struct expression field constructor.
     */
    public static class Field {

        private final Rex key;
        private final Rex value;

        private Field(Rex key, Rex value) {
            this.key = key;
            this.value = value;
        }

        public Rex getKey() {
            return key;
        }

        public Rex getValue() {
            return value;
        }
    }

    private static class Impl extends RexStruct {

        @NotNull
        private final List<Field> fields;

        private Impl(@NotNull List<Field> fields) {
            this.fields = fields;
        }

        @Override
        @NotNull
        public List<Field> getFields() {
            return fields;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexStruct.kt
package org.partiql.plan.rex

/**
 * TODO DOCUMENTATION
 */
public interface RexStruct : Rex {

    public fun getFields(): List<Field>

    override fun getChildren(): Collection<Rex> {
        val children = mutableListOf<Rex>()
        for (field in getFields()) {
            children.add(field.getKey())
            children.add(field.getValue())
        }
        return children
    }

    override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitStruct(this, ctx)

    /**
     * TODO DOCUMENTATION
     */
    public class Field(
        private var key: Rex,
        private var value: Rex,
    ) {
        public fun getKey(): Rex = key
        public fun getValue(): Rex = value
    }
}

/**
 * Default [RexStruct] implementation intended for extension.
 */
internal class RexStructImpl(fields: List<RexStruct.Field>, type: RexType) : RexStruct {

    // DO NOT USE FINAL
    private var _fields = fields
    private var _children: Collection<Rex>? = null
    private var _type = type

    override fun getFields(): List<RexStruct.Field> = _fields

    override fun getType(): RexType = _type

    override fun getChildren(): Collection<Rex> {
        if (_children == null) {
            _children = super.getChildren()
        }
        return _children!!
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RexStruct) return false
        if (_fields != other.getFields()) return false
        return true
    }

    override fun hashCode(): Int = _fields.hashCode()
}
